File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/irpf/ValoresCreditoReclamanteAnterior.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import java.math.BigDecimal;
import java.util.Date;
import java.util.List;

class ValoresCreditoReclamanteAnterior {
    private BigDecimal pagoAnterior;
    private BigDecimal diferencaPrincipalAnterior;
    private Date dataAnterior;
    private BigDecimal numeroCompetencia = BigDecimal.ZERO;

    public ValoresCreditoReclamanteAnterior(BigDecimal pagoAnterior, BigDecimal diferencaPrincipalAnterior, Date dataAnterior) {
        this.pagoAnterior = pagoAnterior;
        this.diferencaPrincipalAnterior = diferencaPrincipalAnterior;
        this.dataAnterior = dataAnterior;
    }

    public BigDecimal getPagoAnterior() {
        return this.pagoAnterior;
    }

    public BigDecimal getDiferencaPrincipalAnterior() {
        return this.diferencaPrincipalAnterior;
    }

    public Date getDataAnterior() {
        return this.dataAnterior;
    }

    public void setPagoAnterior(BigDecimal pagoAnterior) {
        this.pagoAnterior = pagoAnterior;
    }

    public BigDecimal getNumeroCompetencia() {
        return this.numeroCompetencia;
    }

    public void setNumeroCompetencia(BigDecimal numeroCompetencia) {
        this.numeroCompetencia = numeroCompetencia;
    }

    protected static void incluirNoValorAnterior(BigDecimal numeroCompetencias, Date dataPeriodo, List<ValoresCreditoReclamanteAnterior> valoresAnteriores) {
        for (ValoresCreditoReclamanteAnterior vcra : valoresAnteriores) {
            if (!HelperDate.dateEquals(vcra.getDataAnterior(), dataPeriodo)) continue;
            vcra.setNumeroCompetencia(numeroCompetencias);
            break;
        }
    }

    protected static BigDecimal calcularCompetenciasSaldo(BigDecimal totalMeses, List<ValoresCreditoReclamanteAnterior> valoresAnteriores) {
        BigDecimal competenciaTotal = totalMeses;
        for (ValoresCreditoReclamanteAnterior vcra : valoresAnteriores) {
            competenciaTotal = Utils.subtrair(competenciaTotal, vcra.getNumeroCompetencia(), competenciaTotal);
        }
        return competenciaTotal;
    }
}

File: src/lib/pjecalc/core/dominio/calculo/irpf/valores-credito-reclamante-anterior.ts
/**
 * PJe-Calc v2.15.1 — ValoresCreditoReclamanteAnterior
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf.ValoresCreditoReclamanteAnterior
 *
 * Ref Java: pjecalc-fonte/.../calculo/irpf/ValoresCreditoReclamanteAnterior.java (package-private)
 *
 * Registro auxiliar usado pelo `MaquinaDeCalculoDeIrpf` para armazenar
 * valores anteriores de crédito (pago + diferença principal) por data.
 * Inclui `numeroCompetencia` que é preenchido posteriormente.
 */
import Decimal from 'decimal.js';
import { HelperDate } from '../../../base/comum/helper-date';

const ZERO = new Decimal(0);

export class ValoresCreditoReclamanteAnterior {
  private pagoAnterior: Decimal;
  private diferencaPrincipalAnterior: Decimal;
  private dataAnterior: Date;
  private numeroCompetencia: Decimal = ZERO;

  constructor(pagoAnterior: Decimal, diferencaPrincipalAnterior: Decimal, dataAnterior: Date) {
    this.pagoAnterior = pagoAnterior;
    this.diferencaPrincipalAnterior = diferencaPrincipalAnterior;
    this.dataAnterior = dataAnterior;
  }

  getPagoAnterior(): Decimal { return this.pagoAnterior; }
  setPagoAnterior(v: Decimal): void { this.pagoAnterior = v; }

  getDiferencaPrincipalAnterior(): Decimal { return this.diferencaPrincipalAnterior; }
  getDataAnterior(): Date { return this.dataAnterior; }

  getNumeroCompetencia(): Decimal { return this.numeroCompetencia; }
  setNumeroCompetencia(v: Decimal): void { this.numeroCompetencia = v; }

  /** incluirNoValorAnterior (Java linha 48) — localiza pela data e atualiza numero. */
  static incluirNoValorAnterior(
    numeroCompetencias: Decimal,
    dataPeriodo: Date,
    valoresAnteriores: ValoresCreditoReclamanteAnterior[],
  ): void {
    for (const vcra of valoresAnteriores) {
      if (!HelperDate.dateEquals(vcra.getDataAnterior(), dataPeriodo)) continue;
      vcra.setNumeroCompetencia(numeroCompetencias);
      break;
    }
  }

  /** calcularCompetenciasSaldo (Java linha 56) — subtrai todas as competências da lista. */
  static calcularCompetenciasSaldo(
    totalMeses: Decimal,
    valoresAnteriores: ValoresCreditoReclamanteAnterior[],
  ): Decimal {
    let competenciaTotal = totalMeses;
    for (const vcra of valoresAnteriores) {
      competenciaTotal = competenciaTotal.minus(vcra.getNumeroCompetencia());
    }
    return competenciaTotal;
  }
}
